Move metadata server to separate service, create memdriver to test things with pubsub and introduce app discovery endpoint

uncleDecart opened the pull request:
### Ouline

1) It carves out metadata server from zedrouter to separate service (**not created by zedbox**)
2) It creates and invokes metadata server *within* zedrouter and passes http.Handler to LinuxNIReconciler
3) It introduces memdriver for pubsub, which can be used for testing. I don't need to expose internal structures of services, I can just publish things I want. (See msrv_test.go for more)
4) It introduces App Discovery endpoint for metadata server, allowing to discover IP addresses of other App Instances

### Reasoning 

**Refactoring**
Metadata server is a server which is exposed to all AppInstances connected to local Network Instance. It has information about network, can expose cloud-init and also exposes information for PatchEnvelopes. It does relate to networking, however logically it is not exactly part of Zedrouter. In order to reduce number of pubsubs from zedrouter and clean up code I removed it to separate service, which is is a part of zedrouter by composition. Running it as a separate service doesn't make sense, because it'll create additional dependency: metadata server is part of LinuxNIReconciler, so it needs network to run, and since LinuxNIReconciler depends on metadata server to create network, there's kind of circular dependency, we can solve it by communicating via pubsub, but it'll create additional overhead. Also it's gonna be very tricky exposing single metadata server to multiple local network instances :) Also, having metadata server separate from zedrouter allows us to test it without creating network. In order to test metadata server I created memdriver for pubsub, so that I don't need to expose internal structures of services so that I can set them for tests. I just create publications that I know are consumed for specific logic (handler) and publish them :D

**App Discovery endpoint**
This way sidecar application can now IP address of consumer application(s) using this discovery endpoint :) 

### Important
- need https://github.com/lf-edge/eve/pull/3933 to be merged before (cherry picked commit in this PR, will rebase if needed) 

### TODO
- Composing this, I think it makes sense to make a flag persist for Msrv struct, so that for testing we enable those (because it's imperative for memdriver to use persist, because of pubsub logic) and for being part of zedrouter we won't use persist

cc: @deitch @milan-zededa @eriknordmark

 @christoph-zededa  you might be interested in memdriver and msrv_test.go as another way to test EVE services :D 

P.S.
*Pheew* it took me a week to discover this bug above with pubsub :art:

christoph-zededa commented on pkg/pillar/cmd/msrv/handlers.go:
```diff
@@ -524,13 +524,55 @@ func HandleAppCustomBlobs(z *Msrv) func(http.ResponseWriter, *http.Request) {
+		appStatus := srv.lookupAppInstStatusByAppIP(remoteIP)
+		if appStatus == nil {
+			msg := fmt.Sprintf("No AppNetworkStatus for %s", remoteIP.String())
+			srv.Log.Errorf(fmt.Sprintf("HandleAppInstanceDiscovery: %s", msg))
```
`Errorf` does not take a format string?

uncleDecart replied:
oh, yes, you're right, Sprintf is not needed here :)

christoph-zededa commented on pkg/pillar/cmd/msrv/pubsub.go:
```diff
@@ -457,3 +458,47 @@ func (srv *Msrv) handleAppInstDelete(ctxArg interface{}, key string,
+				continue
+			}
+			adapters = append(adapters, AppInstDiscovery{
+				Port:    adapterStatus.IfIdx,
```
What does the user do with the interface index?

uncleDecart replied:
I changed it to Vif. In case where the client application has prior knowledge of the exact physical port on which the server will bind. In this case it will help to find the address client should connect on.

uncleDecart commented:
Rebased to latest master, solved merge conflicts

eriknordmark reviewed (commented):
I see you are also adding AllowToDiscover as part of this PR, but that isn't mentioned in the PR description. Please add.

milan-zededa commented:
I wonder if it would make sense to split the PR, at least between pubsub changes and zedrouter changes.
(request from person infamous for large refactoring PRs :))

deitch commented:
I really so much want this to be a completely standalone service. I don't think we have a service that starts itself and another "sub-service" anywhere else in eve. @uncleDecart explained to me privately that it would be quite difficult to do, so I have to accept it. I do think it is worth expanding on a bit here - not in the PR comments, but in comments right where we run it - why it is run as part of zedrouter and not as a separate service.

This is 9 commits, which probably need to be squashed.

The long list of subscriptions it needs to create further emphasizes the requirement to get pubsub usage simpler, if we can, but that is beyond scope of this PR.

uncleDecart commented on pkg/pillar/go.mod:
```diff
@@ -28,7 +28,7 @@ require (
 	github.com/k8snetworkplumbingwg/network-attachment-definition-client v1.5.0
 	github.com/lf-edge/edge-containers v0.0.0-20240207093504-5dfda0619b80
-	github.com/lf-edge/eve-api/go v0.0.0-20240424223403-4feef58259f3
+	github.com/lf-edge/eve-api/go v0.0.0-20240520195755-0462d38358c9
```
There's a change to introduce flag allow_to_discover to Network Instance I'm afraid :c 

If you can create separate PR bumping EVE API version I can can rebase on master, so you'll get your satisfaction unlike what Rolling Stones sang :D

milan-zededa replied:
Eve already uses eve-api commit https://github.com/lf-edge/eve-api/commit/4feef58259f30f1a57d175af8ed710edd9e3d5d1
I do not see anything other than MTU changes since then: https://github.com/lf-edge/eve-api/commits/main/

uncleDecart replied:
Oh, then I shall remove this immediately!

uncleDecart replied:
But that means, someone bumped EVE API with my changes :c Oh well

milan-zededa replied:
Because you didn't complain :)

uncleDecart commented:
> I see you are also adding AllowToDiscover as part of this PR, but that isn't mentioned in the PR description. Please add.

@eriknordmark added section in PR description, will expand it later a bit 

>I wonder if it would make sense to split the PR, at least between pubsub changes and zedrouter changes.
(request from person infamous for large refactoring PRs :))

That was one of your PRs I remember reviewing 
![image](https://github.com/lf-edge/eve/assets/31950564/61ac9396-93f7-46fe-98af-1775a5c9eb1e)
If you're saying that it's not enough refactoring in my PR, I agree @milan-zededa :D. Thing is that PubSub changes introduce memdriver, which is used in discovery service test. This work is separated into commits, so if you want to focus on particular changes, check things per commit. I think it might actually make sense to leave commits unsquashed. I'll spread comments, licensing and all where it all belongs (last commit as of now) and if we'll think that merging those commits isn't a good idea and squashing zedrouter and pubsub changes would make sense, we can do separate PRs (or I can squash them in this PR :D) 

> The long list of subscriptions it needs to create further emphasizes the requirement to get pubsub usage simpler, if we can, but that is beyond scope of this PR.

Oh well, there's always space for some refactoring :art: , but I agree @deitch. PubSub needs revision, first and foremost, we need to understand the requirements for pubsub 2.0. For instance, what we discussed privately: we want filtering, not different namespaces for subscriptions. I think it's time get this all in one list and see if there's something out there, which will fulfill our requirements or if there should be some more refactoring for PubSub involved. 


> I really so much want this to be a completely standalone service. I don't think we have a service that starts itself and another "sub-service" anywhere else in eve. @uncleDecart explained to me privately that it would be quite difficult to do, so I have to accept it. I do think it is worth expanding on a bit here - not in the PR comments, but in comments right where we run it - why it is run as part of zedrouter and not as a separate service.

Yep, makes sense to add comment explaining why it's running like this. Will do that.

uncleDecart commented:
Yetus is fixed, comments addressed :)

eriknordmark reviewed (changes requested):
I can't tell from the whole diffs, nor from looking at the commit diffs, what changes if any where done to the meta-data server code as part of this since ito shows the files as deleted and new.
I don't know if squashing the two commits [pillar: extract metadata server from zedrouter, and zedrouter: Remove metadata server and use msrv service) would make that be visible, or if you need to make two different commits - one which just does the renames (from cmd/zedrouter to cmd/msrv), and a second commit which does the modifications and additions in cmd/msrv.



Just save the final code so you can compare the the final code doesn't change as a result of doing the commits differently.

If the changes are trivial point me at where in the diffs I can find them.

uncleDecart commented:
@eriknordmark I squashed the [commits](https://github.com/lf-edge/eve/pull/3937/commits/6f4569690c0d9ae0f567bf3eacac9701cc8865db), it helped. Handlers are moved and also I repacked them to be functions instead of different structures which define ServeHTTP, this way handlers are methods of Msrv struct.

milan-zededa commented on pkg/pillar/cmd/msrv/msrv.go:
```diff
@@ -0,0 +1,642 @@
+		msrv.subWwanMetrics,
+		msrv.subDomainStatus,
+		msrv.subAppNetworkStatus,
+
```
extra empty line

uncleDecart replied:
removed

milan-zededa commented on pkg/pillar/cmd/msrv/msrv.go:
```diff
@@ -0,0 +1,642 @@
+	zedcloudCtx := zedcloud.NewContext(msrv.Log, zedcloud.ContextOptions{})
+
+	r.Route("/eve/v1", func(r chi.Router) {
+		r.Get("/network.json", handleNetwork(msrv))
```
These handlers can also be methods of Msrv, and here you can pass `msrv.handleNetwork` etc.

uncleDecart replied:
done

milan-zededa commented on pkg/pillar/cmd/zedrouter/zedrouter.go:
```diff
@@ -198,15 +158,40 @@ func Run(ps *pubsub.PubSub, logger *logrus.Logger, log *base.LogObject, args []s
+	// have to combine all the information published from different instances,
+	// or we can create separate Metadata server and wrap it in a  network
+	// namespace for metadata service so that we won't run into any collisions,
+	// b) bridge LocalNIs to network namespace of metadata server
```
"b)" ?

uncleDecart replied:
rewrote comment a bit :)

eriknordmark reviewed (changes requested):
The squash made it slightly easier to review, but I still have to manually diff the metadata_handlers.go and metatadata.go against msrv/handlers.go and msrv/msrv.go.

As part of this I see some changes burried in this PR which should be discussed as separate commits and/or PRs. One is having sendError() return json instead of text. Another is the patchenvelopes returning a 404 when there are no envelopes instead of a 200 with an empty response. (To be the latter is odd, because the client asked about the list of patch envelops and that resource does exist - the content is an empty list. Which is different than the API endpoint/URL/resource not existing.)

But there can be more - hard to see because the functions appear to be reordered between the old and new files so a diff -iu still has lots of noise. 
Are there other semantic changes? Something about the counters for patch envelopes?

eriknordmark commented on pkg/pillar/cmd/msrv/msrv.go:
```diff
@@ -0,0 +1,639 @@
+// Copyright (c) 2024 Zededa, Inc.
```
Is this derived from zedrouter/metadata.go? If so it makes sense to add 2024 to the existing copyright year.

uncleDecart replied:
Done

eriknordmark commented on pkg/pillar/cmd/msrv/msrv.go:
```diff
@@ -0,0 +1,639 @@
+// Really a constant
+var nilUUID uuid.UUID
+
+// Version is set from the Makefile.
+var Version = "No version specified"
+
```
We've removed this from all existing cmd/*/*.go so don't add it back here. (And I suspect it is not used.)

uncleDecart replied:
Removed

eriknordmark commented on pkg/pillar/cmd/msrv/handlers.go:
```diff
@@ -0,0 +1,719 @@
+// Copyright (c) 2024 Zededa, Inc.
```
This is derived from zedrouter/metadata_handlers.go hence it should add 2024 to the existing copyright years in that file.

uncleDecart replied:
Fixed

eriknordmark commented on pkg/pillar/cmd/msrv/pubsub.go:
```diff
@@ -0,0 +1,506 @@
+package msrv
```
Please add copyright and license lines to new files like this.

uncleDecart replied:
Done

uncleDecart commented:
> As part of this I see some changes burried in this PR which should be discussed as separate commits and/or PRs. One is having sendError() return json instead of text. Another is the patchenvelopes returning a 404 when there are no envelopes instead of a 200 with an empty response. (To be the latter is odd, because the client asked about the list of patch envelops and that resource does exist - the content is an empty list. Which is different than the API endpoint/URL/resource not existing.)

Agreed, removed it. Functions are same from zedrouter. There was also mixup in commits, cleaned it up.

eriknordmark commented:
> > As part of this I see some changes burried in this PR which should be discussed as separate commits and/or PRs. One is having sendError() return json instead of text. Another is the patchenvelopes returning a 404 when there are no envelopes instead of a 200 with an empty response. (To be the latter is odd, because the client asked about the list of patch envelops and that resource does exist - the content is an empty list. Which is different than the API endpoint/URL/resource not existing.)
> 
> Agreed, removed it. Functions are same from zedrouter. There was also mixup in commits, cleaned it up.

Did you check that there aren't any other fixes/changes burried in those two files? (It is too hard to eye-ball looking for things like that.)

uncleDecart commented:
>>> As part of this I see some changes burried in this PR which should be discussed as separate commits and/or PRs. One is having sendError() return json instead of text. Another is the patchenvelopes returning a 404 when there are no envelopes instead of a 200 with an empty response. (To be the latter is odd, because the client asked about the list of patch envelops and that resource does exist - the content is an empty list. Which is different than the API endpoint/URL/resource not existing.)

>> Agreed, removed it. Functions are same from zedrouter. There was also mixup in commits, cleaned it up.

> Did you check that there aren't any other fixes/changes burried in those two files? (It is too hard to eye-ball looking for things like that.)

Yes, compared handlers and they're the same.

eriknordmark reviewed (approved):
LGTM
